agent/consul/multilimiter/multilimiter.go

```diff
@@ -145,14 +155,45 @@ func (m *MultiLimiter) Allow(e LimitedEntity) bool {
 			config = prefixConfig
 		}
 	}
-
 	limiter := &Limiter{limiter: rate.NewLimiter(config.Rate, config.Burst)}
-	limiter.lastAccess.Store(now)
-	tree, _, _ := limiters.Insert(e.Key(), limiter)
-	m.limiters.Store(tree)
+	limiter.lastAccess.Store(unixNow)
+	m.LimiterCh <- &LimiterWithKey{l: limiter, k: e.Key(), t: now}
 	return limiter.limiter.Allow()
 }
 
+func (m *MultiLimiter) runStore(ctx context.Context) {
+	var txn *radix.Txn
+	waiter := time.NewTimer(10 * time.Millisecond)
+	defer waiter.Stop()
+	for {
+		select {
+		case <-waiter.C:
+			if txn != nil {
+				tree := txn.Commit()
+				m.limiters.Store(tree)
+			}
+		case lk := <-m.LimiterCh:
+			limiters := m.limiters.Load()
+			if txn == nil {
+				txn = limiters.Txn()
+			}
+			v, ok := txn.Get(lk.k)
+			if !ok {
+				txn.Insert(lk.k, lk.l)
+			} else {
+				switch l := v.(type) {
+				case *Limiter:
+					l.lastAccess.Store(lk.t.Unix())
+					l.limiter.AllowN(lk.t, 1)
+				}
```

Comment:
No biggie, but if `v` can only ever be a `*Limiter` we can use a more straightforward type-assertion:

```suggestion
				if l, ok := v.(*Limiter); ok {
					l.lastAccess.Store(lk.t.Unix())
					l.limiter.AllowN(lk.t, 1)
				}
```